ogre2/src/Ogre2Capsule.cc

```diff
@@ -0,0 +1,170 @@
+
+//////////////////////////////////////////////////
+Ogre2Capsule::Ogre2Capsule()
+ : dataPtr(new Ogre2CapsulePrivate)
+{
+}
+
+//////////////////////////////////////////////////
+Ogre2Capsule::~Ogre2Capsule()
+{
+  // no ops
+}
+
+//////////////////////////////////////////////////
+Ogre::MovableObject *Ogre2Capsule::OgreObject() const
+{
+  return this->dataPtr->ogreMesh->OgreObject();
+}
+
+//////////////////////////////////////////////////
+void Ogre2Capsule::PreRender()
+{
+  if (this->capsuleDirty)
+  {
+    this->Create();
+    this->capsuleDirty = false;
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2Capsule::Init()
+{
+  this->Create();
+}
+
+//////////////////////////////////////////////////
+void Ogre2Capsule::Destroy()
+{
+  if (!this->Scene())
+    return;
+
+  if (this->dataPtr->ogreMesh)
+  {
+    this->dataPtr->ogreMesh->Destroy();
+    this->dataPtr->ogreMesh.reset();
+  }
+
+  if (this->dataPtr->material && this->Scene())
+  {
+    this->Scene()->DestroyMaterial(this->dataPtr->material);
+    this->dataPtr->material.reset();
+  }
+}
+
+////////////////////////////////////////////////
+void Ogre2Capsule::Create()
+{
+  common::MeshManager *meshMgr = common::MeshManager::Instance();
+  std::string capsuleMeshName = this->Name() + "_capsule_mesh"
+    + "_" + std::to_string(this->radius)
+    + "_" + std::to_string(this->length);
+  if (!meshMgr->HasMesh(capsuleMeshName))
+  {
+    meshMgr->CreateCapsule(capsuleMeshName, this->radius, this->length, 12, 32);
+    MeshDescriptor meshDescriptor;
+    meshDescriptor.mesh = meshMgr->MeshByName(capsuleMeshName);
+    if (meshDescriptor.mesh != nullptr)
+    {
+      auto visual = std::dynamic_pointer_cast<Ogre2Visual>(this->Parent());
+
+      // clear geom if needed
+      if (this->dataPtr->ogreMesh)
+      {
+        if (visual)
+        {
+          visual->RemoveGeometry(
+              std::dynamic_pointer_cast<Geometry>(shared_from_this()));
+        }
+        this->dataPtr->ogreMesh->Destroy();
+      }
+      this->dataPtr->ogreMesh =
+        std::dynamic_pointer_cast<Ogre2Mesh>(
+          this->Scene()->CreateMesh(meshDescriptor));
+      if (this->dataPtr->material != nullptr)
+      {
+        this->dataPtr->ogreMesh->SetMaterial(this->dataPtr->material, false);
+      }
+      if (visual)
+      {
+        visual->AddGeometry(
+            std::dynamic_pointer_cast<Geometry>(shared_from_this()));
+      }
+    }
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2Capsule::SetMaterial(MaterialPtr _material, bool _unique)
+{
+  _material = (_unique) ? _material->Clone() : _material;
+
+  Ogre2MaterialPtr derived =
+      std::dynamic_pointer_cast<Ogre2Material>(_material);
+
+  if (!derived)
+  {
+    ignerr << "Cannot assign material created by another render-engine"
+        << std::endl;
+
+    return;
+  }
+
+  // Set material for the underlying dynamic renderable
+  this->dataPtr->ogreMesh->SetMaterial(_material, false);
```

Comment:
Should this `ogreMesh->SetMaterial` call be in `SetMaterialImpl`? Otherwise if another function calls that it won't have any effect.

In fact, I'm not sure what's the advantage of having a separate `SetMaterialImpl` function, it looks to me that the entire logic could be inside `SetMaterial`.